### PR TITLE
Router parameters

### PR DESCRIPTION
**Pull request by jwx**
# Pull Request

## 📖 Description

Parameters for the convention-based router. Preliminary example:

`app`

```html
<template>
  <div style="padding: 20px;">
    <nav>
      <ul>
        <li><a href="authors+about">Authors</a></li>
        <li><a href="books+about">Books</a></li>
        <li><a href="about">About</a></li>
      </ul>
    </nav>
    <au-viewport name="lists" used-by="authors,books" default="authors"></au-viewport>
    <au-viewport name="content" default="about"></au-viewport>
  </div>
</template>
```

```ts
export class App {
  constructor(private router: Router) {
    this.router.activate();
  }
}
```

`authors` (and `books` is the same)

```html
<template>
<h3>Authors</h3>
<ul>
  <li repeat.for="author of authors">
    <a href="author=${author.id}">${author.name}</a>
    <ul><li repeat.for="book of author.books">${book.title}</li></ul>
  </li>
</ul>
</template>
```

```ts
export class Authors {
  constructor(private authorsRepository: AuthorsRepository) { }

  get authors() { return this.authorsRepository.authors(); }
}
```

`authors` (and `books` is the same)

```html
<template>
<h3>${author.name}</h3>
<div>Born: ${author.born}</div>
<div>Books:
  <ul>
    <li repeat.for="book of author.books"><a href="book=${book.id}">${book.title}</a></li>
  </ul>
</div>
</template>
```

```ts
export class Author {
  public static parameters = ['id'];

  public author = {};
  constructor(private authorsRepository: AuthorsRepository) { }

  public enter(parameters) {
    if (parameters.id) {
      this.author = this.authorsRepository.author(+parameters.id);
    }
  }
}
```

### 🎫 Issues

Related to https://github.com/aurelia/aurelia/issues/307

## 👩‍💻 Reviewer Notes

- Comments from @fkleuver regarding code quality can be found [here](https://github.com/aurelia/aurelia/pull/354) (in a now discarded PR). Some of those are addressed, some are still TODO
- Test application can be found in packages/router/test/e2e/nav. (Select Lobby followed by Contacts.)

## 📑 Test Plan

- Added ~10 unit tests to verify parameter parsing and behavior
- Added a test app (intended for inclusion in the e2e tests asap)

## ⏭ Next Steps

- Add an extensibility point for customizing the convention-based url syntax
- Clean up / improve typings, etc

### CODE REVIEW

File: packages/router/src/scope.ts [hunk 1]
```diff
@@ -18,15 +18,30 @@ export interface IFindViewportsResult {
 }
 
 export class Scope {
+  public element: Element;
+  public parent: Scope;
+
   public viewport: Viewport;
 
   public children: Scope[] = [];
   public viewports: Object = {};
```
Comment (fkleuver) [inline packages/router/src/scope.ts]:
```suggestion
  public viewports: Record<string, Viewport>= {};
```

File: packages/router/src/scope.ts [hunk 2]
```diff
@@ -18,15 +18,30 @@ export interface IFindViewportsResult {
...
 
+  private readonly router: Router;
+
   private scopeViewportParts: Object = {};
```
Comment (fkleuver) [inline packages/router/src/scope.ts]:
```suggestion
  private scopeViewportParts: Record<string, Viewport>= {};
```

File: packages/router/src/scope.ts [hunk 3]
```diff
@@ -18,15 +18,30 @@ export interface IFindViewportsResult {
...
   private availableViewports: Object;
```
Comment (fkleuver) [inline packages/router/src/scope.ts]:
```suggestion
  private availableViewports: Record<string, Viewport>;
```

File: packages/router/src/router.ts [hunk 4]
```diff
@@ -49,10 +51,16 @@ export class Router {
   public historyBrowser: HistoryBrowser;
   public linkHandler: LinkHandler;
 
+  public navs: Object = {};
```
Comment (fkleuver) [inline packages/router/src/router.ts]:
```suggestion
  public navs: Record<string, Nav> = {};
```

File: packages/router/src/history-browser.ts [hunk 5]
```diff
@@ -3,6 +3,9 @@ export interface IHistoryEntry {
   fullStatePath: string;
   index?: number;
   title?: string;
+  query?: string;
+  parameters?: Object;
```
Comment (fkleuver) [inline packages/router/src/history-browser.ts]:
```suggestion
  parameters?: Record<string, unknown>;
```

File: packages/router/src/history-browser.ts [hunk 6]
```diff
@@ -3,6 +3,9 @@ export interface IHistoryEntry {
...
+  parameterList?: string[];
   data?: Object;
```
Comment (fkleuver) [inline packages/router/src/history-browser.ts]:
```suggestion
  data?: Record<string, unknown>;
```

File: packages/router/src/history-browser.ts [hunk 7]
```diff
@@ -70,7 +73,7 @@ export class HistoryBrowser {
   public goto(path: string, title?: string, data?: Object): void {
```
Comment (fkleuver) [inline packages/router/src/history-browser.ts]:
```suggestion
  public goto(path: string, title?: string, data?: Record<string, unknown>): void {
```

File: packages/router/src/history-browser.ts [hunk 8]
```diff
@@ -286,6 +299,13 @@ export class HistoryBrowser {
     this.pathChanged();
   }
 
+  private getSearch(): string {
+    const hash = this.location.hash.substr(1) || '';
```
Comment (fkleuver) [inline packages/router/src/history-browser.ts]:
```suggestion
    const hash = this.location.hash.length > 1 ? this.location.hash.slice(1) : '';
```

File: packages/router/src/history-browser.ts [hunk 9]
```diff
@@ -286,6 +299,13 @@ export class HistoryBrowser {
...
+    const hashSearches = hash.split('?');
+    hashSearches.shift();
+    return hashSearches.shift() || '';
```
Comment (fkleuver) [inline packages/router/src/history-browser.ts]:
```suggestion
    return hashSearches.length > 0 ? hashSearches.shift() : '';
```

File: packages/router/src/nav-route.ts [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+  component: string | ICustomElementType;
+}
+
+export class NavRoute {
+  public nav: Nav;
+  public components: string | ICustomElementType | IViewportComponent;
+  public title: string;
+  public link?: string;
+  public children?: NavRoute[];
+  public meta?: Object;
```
Comment (fkleuver) [inline packages/router/src/nav-route.ts]:
```suggestion
  public meta?: Record<string, string>;
```

File: packages/router/src/nav.ts [hunk 11]
```diff
@@ -0,0 +1,40 @@
+import { CustomElementResource, ICustomElementType } from '@aurelia/runtime';
+import { NavRoute } from './nav-route';
+import { Router } from './router';
+
+export interface INavRoute {
+  components: string | ICustomElementType | Object;
+  link?: string;
+  title: string;
+  children?: NavRoute[];
+  meta?: Object;
```
Comment (fkleuver) [inline packages/router/src/nav.ts]:
```suggestion
  meta?: Record<string, string>;
```

File: packages/router/src/parser.ts [hunk 12]
```diff
@@ -0,0 +1,61 @@
+export interface IParsedQuery {
+  parameters: Object;
```
Comment (fkleuver) [inline packages/router/src/parser.ts]:
```suggestion
  parameters: Record<string, string>;
```

File: packages/router/src/parser.ts [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+  list: string[];
+}
+
+export interface IMergedParameters {
+  parameters: Object;
```
Comment (fkleuver) [inline packages/router/src/parser.ts]:
```suggestion
  parameters: Record<string, string>;
```

File: packages/router/src/parser.ts [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+  list: string[];
+  merged: string[] | Object;
```
Comment (fkleuver) [inline packages/router/src/parser.ts]:
```suggestion
  merged: string[] | Record<string, string>;
```

File: packages/router/src/router.ts [hunk 15]
```diff
@@ -134,6 +155,10 @@ export class Router {
       }
     }
 
+    const parsedQuery: IParsedQuery = parseQuery(instruction.query);
+    instruction.parameters = parsedQuery.parameters;
+    instruction.parameterList = parsedQuery.list;
+
     let title;
     let views: Object;
```
Comment (fkleuver) [inline packages/router/src/router.ts]:
```suggestion
    let views: Record<string, ViewPort>;
```

**Comment by fkleuver** [comment]
Alright I did a couple typing fixes. It's a little unfortunate the test coverage went down so much but as you said the `nav` things are neither used nor tested properly at the moment. I'm (sort of) okay with leaving them in until there is a `@aurelia/router-html` package of sorts that can be home to HTML specific stuff. I think that's where nav stuff can live.

Overall this is a pretty small PR that just looks big due to the test app. The lifecycle hooks are now properly wired up in the viewport as we discussed and I'm quite glad that they're there now.

I don't want to get too nit picky over the various potential edge cases and whatnot right now, so I think in that context it's a good addition of a feature.
There is a test app that still needs to be hooked up in automated e2e tests (potentially also to be included in coverage calculations) and you said you would work more on code quality as well, making it easier for me to pick things up / integrate when time comes.

So with all the above in mind, I think this is good to merge. Let's get it in @EisenbergEffect

**Comment by fkleuver** [comment]
Removed WIP from title as per our discussion since only typings were left and I fixed those

**Comment by EisenbergEffect** [comment]
Let's try to bring the code coverage up on this. Project's total coverage has slowly dropped by approx. 2% recently.
